### PR TITLE
Store label in SharedAllocationRecord

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/2029. Supersedes #4615.

### CODE REVIEW

**Comment by masterleinad** [comment]
Storing the labels in SharedAllocationRecord additionally, seems to also fix the broken tests in `TestEventCorrcetness.hpp`.

File: core/src/impl/Kokkos_SharedAlloc.hpp [hunk 1]
```diff
@@ -149,7 +150,7 @@ class SharedAllocationRecord<void, void> {
       SharedAllocationRecord* arg_root,
 #endif
       SharedAllocationHeader* arg_alloc_ptr, size_t arg_alloc_size,
-      function_type arg_dealloc);
+      function_type arg_dealloc, const std::string& label);
```
Comment (dalg24) [inline core/src/impl/Kokkos_SharedAlloc.hpp:153]: Consider taking it by value and moving it.

**Review by dalg24 — APPROVED** [review]
Pleas expand the description.  The original issue consider multiple options.  Make it clear which one you implemented.

File: core/src/impl/Kokkos_SharedAlloc.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_SharedAlloc.hpp:153]: In most (all?) cases, we obtain the label as a `const std::string&` already so I opted for passing by `const std::string&` here, too. If there are other good reasons to rather pass by value and move, I am happy to adapt, of course.

**Comment by masterleinad** [comment]
This implements https://github.com/kokkos/kokkos/issues/2029#issuecomment-475230552, i.e we are storing the label additionally/redundantly in `SharedAllocationRecord`.

**Review by crtrott — APPROVED** [review]

File: core/src/impl/Kokkos_SharedAlloc.hpp (hunk 1 above)
Comment (PhilMiller) [reply core/src/impl/Kokkos_SharedAlloc.hpp:153]: As written, there's guaranteed to be a copy into the ultimate `m_label` member.

In the common case where the label is coming from some temporary object, Damien's suggestion will allow the compiler to do elision, and construct the temporary directly into the function argument, which will then get inexpensively moved from.

Comment (masterleinad) [reply core/src/impl/Kokkos_SharedAlloc.hpp:153]: I agree with that but we never pass a temporary object here.
Anyway, if anyone requests changes, I will change.

**Pull request merged**